crypto/evp/e_aes.c

```diff
@@ -390,22 +390,42 @@ static int aesni_xts_init_key(EVP_CIPHER_CTX *ctx, const unsigned char *key,
         return 1;
 
     if (key) {
+        /* The key is two half length keys in reality */
+        const int bits = EVP_CIPHER_CTX_key_length(ctx) * 8 / 2;
```

Comment:
You could also define `bytes` here, in order to avoid all the divisions by 8 below.

```C
        const int bits = EVP_CIPHER_CTX_key_length(ctx) * (8 / 2);
        const int bytes = bits / 8;
```

Reply:
And why don't you say,
bytes = EVP_CIPHER_CTX_key_length(ctx) / 2;
bits = bytes * 8;

Reply:
Answer from a pedantic mathematician: your proposal  is equivalent, provided the key length is even ;-)

Reply:
Serious answer: +1

Reply:
Should we check that key length is even?

Reply:
at least the memcmp below relies on that fact.

Reply:
I would guess the key length is always a power of two, but I don’t know whether this has been checked elsewhere. @paulidale?

Reply:
This is for AES, so yes it is always even.  It isn't always a power of two (AES-192).

There are no divisions or multiplications, just shifts.

Reply:
I use @bernd-edlinger 's suggestion for bits and bytes.

Reply:
I was wrong, AES-XTS is not defined using AES-192, so the key will be a power of two after all (either 256 or 512 bits).